clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java

```diff
@@ -0,0 +1,354 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * A LinkedHashSet which is more memory-efficient than the standard implementation.
+ *
+ * This set preserves the order of insertion.  The order of iteration will always be
+ * the order of insertion.
+ *
+ * This collection requires previous and next indexes to be embedded into each
+ * element.  Using array indices rather than pointers saves space on large heaps
+ * where pointer compression is not in use.  It also reduces the amount of time
+ * the garbage collector has to spend chasing pointers.
```

Comment:
I think the optimization of using array indices instead of pointers is a bit questionable without some benchmarks. Heaps larger than 32 GB are rarely (or never) used in Kafka. And having to go via the array has some cost as well.

Reply:
There are other benefits besides reducing the pointer size.  When you use array indices rather than pointers, the garbage collector needs to do less work chasing pointers.  See https://issues.apache.org/jira/secure/attachment/12701400/BlocksMap%20redesign.pdf .

Excerpt:
According to an Oracle engineer, large heaps with reference dense objects in old gen with frequently mutating references is brutally hard on GC.When a reference in an old gen object is mutated, the object’s “card page” is marked as dirty. During young gen collection all references in dirty old gen card pages are used as roots for determining reachability of young gen objects.

The [HDFS] block data-structure mutates by necessity, but it does so in a non-GC friendly
manner. Report processing inserts a delimiter into the storage’s doubly linked list,
moves reported blocks to the head of the storage’s list, then uses the delimiter to
determine excess blocks for invalidation. The updating of so many references creates
intense pressure on GC.

One reason is young gen maintains a tenuring threshold equating to how conservatively
it will promote young gen objects into old gen. The threshold drops relative to the rate
of garbage creation and dirtying of old gen cards. The young collector may resort to
prematurely promoting objects into old gen when it becomes overrun by spending too
much time collecting. CMS is forced to cleanup when the old gen occupancy threshold
is exceeded. The prematurely promoted objects lead to excessive fragmentation of old
gen.

We can reduce abusive GC behavior by reducing the mutation of references in old gen.
Unlike references, updating primitives (ints, longs, etc) does not mark an old gen page dirty.  It does not incur a penalty to young gen collection.
```